mlflow/models/evaluation/default_evaluator.py

```diff
@@ -614,7 +662,53 @@ def plot_pr_curve():
 
         self._log_pandas_df_artifact(per_class_metrics_collection_df, "per_class_metrics")
 
-    def _evaluate_custom_metrics(self):
+    def _log_custom_metric_artifact(
+        self, artifact_name, raw_artifact, custom_metric_index, custom_metric_name
+    ):
+        if isinstance(raw_artifact, EvaluationArtifact):
+            return raw_artifact
+
+        exception_header = _get_custom_metric_exception_header(
+            custom_metric_index, custom_metric_name
+        )
+
+        inferred_from_path, inferred_type, inferred_ext = _infer_artifact_type_and_ext(
+            artifact_name, raw_artifact, exception_header
+        )
+        artifact_file_name = _gen_log_key(artifact_name, self.dataset_name) + inferred_ext
+        artifact_file_local_path = self.temp_dir.path(artifact_file_name)
+
+        if inferred_from_path:
+            shutil.copyfile(raw_artifact, artifact_file_local_path)
+        elif inferred_type is CsvEvaluationArtifact:
+            raw_artifact.to_csv(artifact_file_local_path, index=False)
+        elif inferred_type is NumpyEvaluationArtifact:
+            np.save(artifact_file_local_path, raw_artifact, allow_pickle=False)
+        elif inferred_type is ImageEvaluationArtifact:
+            raw_artifact.savefig(artifact_file_local_path)
+        else:
+            # try serializing as JSON.
+            try:
+                json.dump(raw_artifact, open(artifact_file_local_path, "w"))
+            except TypeError:
+                raise MlflowException(
+                    f"{exception_header} produced an unsupported artifact '{artifact_name}' with "
+                    f"type '{type(raw_artifact)}'. Supported object types for artifacts are:"
+                    f"- A string uri representing the file path to the artifact. MLflow"
+                    f"  will infer the type of the artifact based on the file extension."
+                    f"- EvaluationArtifact type objects. i.e. ImageEvaluationArtifact."
+                    f"- Pandas DataFrame. This will be resolved as a CSV artifact."
+                    f"- Numpy array. This will be saved as a .npy artifact."
+                    f"- Matplotlib Figure. This will be saved as an image artifact."
+                    f"- Any JSON serializable object. This will be saved as a JSON file."
+                )
+
```

Comment:
Why is JSON our fallback case? Can't we save the string representation of the object as a text file? This seems like a better catch-all.

Reply:
JSON is used as fallback because I think it is more structured and standardized than using string representation. For example, if a user have an arbitrary class that they want to serialize. They can either opt to provide an implementation to `__repr__` and define their own format (which most of the time I think it's a dictionary that is JSON serializable anyways), or they can instead take a more standardized and more established approach of making their class [JSON serializable](https://pythontic.com/serialization/json/introduction). 

I think what we could do is that we first try to JSON serialize it, if it fails, we save the string representation. But imo this is just passing down the problem. Because if their class doesn't have an implementation to `__repr__` they'll see that their object is "serialized" to something like `<__main__.some.package.TheirObject object at 0x7fb7b08ab290>`.

Using string is a better catch-all in the sense that it won't raise an exception when the user doesn't provide a way to serialize their object. I think if we want to be verbose and let them know immediately that their class isn't serializable, then we should stick with JSON. But if we want to be more accommodating, having string as an option could be a good idea.

Reply:
I'd rather we be flexible than rigid with what users can log. If the object is of a string type, can we write it out as txt. If the object is anything else, can we just pickle it? In either case, let's print a warning.